core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -440,7 +440,7 @@ private RexNode simplifyIsPredicate(SqlKind kind, RexNode a) {
       }
       if (kind == pred.kind) {
         return rexBuilder.makeLiteral(true);
-      } else {
+      } else if (kind.negate() == pred.kind) {
         return rexBuilder.makeLiteral(false);
```

Comment:
I wonder if this `else` branch is ever taken.

@kgyrtkirk , you seem to recently introduce that code. Would you please provide a test case for that branch?
All the existing tests pass if I remove `else return false` branch.

Reply:
well,the "or(isNull(vBool()),isNull(vBool()))" will run to the else branch.

Reply:
And it will simplify to isNull(vBool()).

Reply:
@pengzhiwei2018 , what I want to do is as following:
1) comment "else" branch altogether
2) find a case what would lack simplification

`or(isNull(vBool()),isNull(vBool()))` is simplified to `isNull(vBool())` even without that `else` branch.

Alternative option is to just remove the `else` branch for now.

As you see,  the `else` branch causes ERRORs (wrong results after simplification), and the branch was never tested. That is why I want either to make the branch tested (to ensure it does whatever author supposed for it to do) or remove the branch at all (to avoid further invalid simplifications like 2527).

Reply:
It seems that RexUtil#composeDisjunction and RexUtl#composeConjunction wil merge the same RexNodes in "OR" and "AND".So this "else-branch" is no need here.I would like to remove this branch.

Reply:
I agree that it was incorrect without the check added in this patch.
About removing it: I think it will be probably used in: `and(isNull(vBool()), or(isNotNull(vBool()), vBool(2)))`
I've just checked it...its needed; I think this the current patch is right

Reply:
@kgyrtkirk , I suggest to do the following: remove the branch for now, then you (or someone else) might come up with a testcase when else is required.

The testcase you mention `and(isNull(vBool()), or(isNotNull(vBool()), vBool(2)))` does not enter `else if (kind.negate() == pred.kind) {` branch.

Reply:
it does enter that if...the following test passed with the else; and now the simplification doesn't happen...

     @Test public void x() {
          checkSimplify2(
               and( isNull(vBool()), or(isNotNull(vBool()),vBool(2))),
               "AND(IS NULL(?0.bool0), OR(IS NOT NULL(?0.bool0), ?0.bool2))",
               "AND(IS NULL(?0.bool0), ?0.bool2)");
          }